src/server/models/Player.js

```diff
@@ -86,33 +130,119 @@ export default class Player extends Model {
+      }
+      if (eventMap.get(card) <= 0) {
+        eventMap.delete(card);
+      }
+    }
+  }
+
+  addPersistentEffect(eventName, card) {
+    if (!this.persistentEffects.has(eventName)) {
+      this.persistentEffects.set(eventName, new Map());
+    }
+    const eventMap = this.persistentEffects.get(eventName);
+    if (!eventMap.has(card)) {
+      eventMap.set(card, 0);
+    }
+    eventMap.set(card, eventMap.get(card) + 1);
+  }
+
+  removePersistentEffect(eventName, card) {
+    if (this.persistentEffects.has(eventName)) {
+      const eventMap = this.persistentEffects.get(eventName);
+      if (eventMap.has(card)) {
+        eventMap.set(card, eventMap.get(card) - 1);
+      }
+      if (eventMap.get(card) <= 0) {
+        eventMap.delete(card);
       }
-      handledReactions.add(reaction);
-      reactions = this.hand.filter(card => !handledReactions.has(card) && card.shouldReactTo(event, this, player, ...args));
+    }
+  }
 
-      handled = await reaction.reactTo(event, this, player, ...args) || handled;
+  async handleTriggers(eventName, args, otherCardsToCheck) {
+    const event = new Event(eventName, this, this.game, args);
+    await this.handleOwnTriggers(event, otherCardsToCheck);
+    await this.forEachOtherPlayer(player => player.handleOwnTriggers(event, otherCardsToCheck));
+    return event;
+  }
+
+  async handleOwnTriggers(event, otherCardsToCheck = []) {
+    const handledCards = new Set();
+    const cards = {
+      optional: {},
+      mandatory: {},
+    };
+    let optionalCardsLength = 0;
+    let mandatoryCardsLength = 0;
+    const refreshCards = () => {
+      [
+        { title: 'hand', getAllCards: () => [...this.hand] },
+        { title: 'playArea', getAllCards: () => [...this.playArea] },
+        { title: 'reserve', getAllCards: () => (this.mats.tavern ? [...this.mats.tavern] : []) },
+        {
+          title: 'persistent',
+          getAllCards: () => (
+            this.persistentEffects.has(event.name) ?
+              [...this.persistentEffects.get(event.name).entries()].reduce((all, [card, count]) => [...all, ...[...new Array(count)].map(() => card)], []) :
+              []
+          ),
+        },
+        { title: 'other', getAllCards: () => otherCardsToCheck },
+      ].forEach(({ title, getAllCards }) => {
+        cards.optional[title] = getAllCards().filter(card => !handledCards.has(card) && card.canTriggerOn(event, this, title === 'persistent' ? 'persistent' : 'normal'));
+        cards.mandatory[title] = getAllCards().filter(card => !handledCards.has(card) && card.willTriggerOn(event, this, title === 'persistent' ? 'persistent' : 'normal'));
+      });
+      optionalCardsLength = Object.values(cards.optional).reduce((sum, list) => sum + list.length, 0);
+      mandatoryCardsLength = Object.values(cards.mandatory).reduce((sum, list) => sum + list.length, 0);
+    };
+    refreshCards();
+
+    while (optionalCardsLength > 0 || mandatoryCardsLength > 1) {
```

Comment:
probably need mandatoryCardsLength > 0 instead of 1

Reply:
no, you don't get the option to select the last mandatory card
it just happens automatically

Reply:
this way you don't have to click something to make a duration trigger when there is only one there

Reply:
That makes sense, forgot about the automatic implementation